File: src/utils/stats.lua
#!/usr/bin/env lua
-- Calculate summary statistics for a series of numbers

package.path = package.path .. ";utils/?.lua"

local M = require "mean"
local P = require "percentile"

local numbers = {}

for number in io.lines() do
	numbers[#numbers+1] = tonumber(number)
end

if #numbers == 0 then os.exit(1) end

local P0   = P(  0, numbers)
local P10  = P( 10, numbers)
local P25  = P( 25, numbers)
local P50  = P( 50, numbers)
local P75  = P( 75, numbers)
local P90  = P( 90, numbers)
local P100 = P(100, numbers)

print "Min,P10,P25,Median,P75,P90,Max,P75-P25,P90-P10,Max-Min,Mean ± RSD"
print(string.format(string.rep("%.3f,", 10) .. "%.3f ± %.2f %%",
	P0,				-- minimum
	P10,			-- 10th percentile
	P25,			-- 25th percentile / first quartile
	P50,			-- median / second quartile
	P75,			-- 75th percentile / third quartile
	P90,			-- 90th percentile
	P100,			-- maximum
	P75-P25,		-- interquartile range
	P90-P10,		-- central 80 % range
	P100-P0,		-- total range,
	M(numbers)		-- mean ± relative standard deviation in percent
))

File: utils/stats.py
#!/usr/bin/env python3

import argparse
import statistics
import sys

import tabulate as T


def mean_rsd(numbers):
    """
    Calculate mean and relative standard deviation (RSD) of a series of numbers
    """
    mean = statistics.fmean(numbers)
    sd = statistics.stdev(numbers)
    rsd = 100 * sd / abs(mean)
    return mean, rsd


def quartiles(numbers):
    return statistics.quantiles(numbers, n=4, method="inclusive")


def deciles(numbers):
    return statistics.quantiles(numbers, n=10, method="inclusive")


def print_stats(numbers, tabulate=True):
    Q = quartiles(numbers)
    D = deciles(numbers)

    headers = [
        "Min", "P10", "P25", "Median", "P75", "P90", "Max",
        "P75-P25", "P90-P10", "Max-Min", "Mean ± RSD"
    ]

    stats = [
        min(numbers),                 # Minimum
        D[0],                         # 10th percentile / 1st decile
        Q[0],                         # 25th percentile / 1st quartile
        Q[1],                         # 50th percentile / 2nd quartile / median
        Q[-1],                        # 75th percentile / 3rd quartile
        D[-1],                        # 90th percentile / 9th decile
        max(numbers),                 # Maximum
        Q[-1] - Q[0],                 # Interquartile range
        D[-1] - D[0],                 # Interdecile range
        max(numbers) - min(numbers)   # Total range
    ]

    if tabulate:
        stats = [round(stat, 2) for stat in stats]
        # Append mean ± relative standard deviation in percent
        stats.append("{:.2f} ± {:.2f} %".format(*mean_rsd(numbers)))
        print(T.tabulate([stats], headers=headers, tablefmt="pretty"))
    else:
        print(",".join(headers))
        print(",".join(map(lambda stat: f"{stat:.2f}", stats)), end='')
        # Append mean ± relative standard deviation in percent
        print(",{:.2f} ± {:.2f} %".format(*mean_rsd(numbers)))



if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--tabulate", action="store_true", help="tabulate output", required=False)
    args = parser.parse_args()

    numbers = []
    for line in sys.stdin:
        numbers.append(float(line))

    print_stats(numbers, args.tabulate)
